numpy/_core/tests/test_multiarray.py

```diff
@@ -5012,7 +5013,19 @@ def test_basic(self):
             self._clip_type(
                 'uint', 1024, 0, 0, inplace=inplace)
             self._clip_type(
-                'uint', 1024, -120, 100, inplace=inplace, expected_min=0)
+                'uint', 1024, 10, 100, inplace=inplace)
+
+    @pytest.mark.parametrize("inplace", [False, True])
+    def test_int_range_error(self, inplace):
+        # E.g. clipping uint with negative integers fails to promote
+        # (changed with NEP 50 and may be adaptable)
+        # Similar to last check in `test_basic`
+        x = (np.random.random(1000) * 255).astype("uint8")
+        with pytest.raises(OverflowError):
+            x.clip(-1, 10, out=x if inplace else None)
```

Comment:
Hmm, should we change `clip` as well to use `resolve_descriptors_raw`?

Reply:
You will have to do a 3 way (you have to check two arguments) dispatching stuff which will be seems like a lot of trouble for a niche feature.  So it seems like a lot of trouble for something very niche?  It isn't even clear to me that the test is testing anything we expect users to do.
(I may be wrong, but I expect more floating point uses and this type of clip call is only relevant when you don't know the input integer dtype and also do not want to cast anyway.)

Reply:
`clip` goes through python code in `_methods.py` that already special-cases `min` and `max` values: https://github.com/numpy/numpy/blob/d885b0b546cbe2ab8401622c8d4db984ec5ea31b/numpy/_core/_methods.py#L92-L101

I made a PR with a possible fix -- https://github.com/seberg/numpy/pull/48 -- though at some level that just brings up the next question: if `clip`, why not also `np.maximum` and `np.minimum` (one-sided only, i.e., `np.minimum(int_array, py_int)` should error if `py_int` is less than the minimum possible given the dtype, but just pass through `int_array` if it is larger than the maximum possible).

In that respect, maybe editing the python function is the right call, since that already special-cases `None` as well.

Reply:
Thanks, I was tempted to make it a follow-up if there is an actual complaint.  Do you really think it is important?

Probably only a theoretical caveat, but all of these function (for clip, I think it is a bad choice), return the result type of mixing everything.  That is slightly surprising to get something that can't hold all values.  Although, I have to admit, it isn't really problematic in practice.

Reply:
After I wrote my change, I must admit I became unsure, especially after having to adjust the tests for numpy ints to not do in-place -- as you say, this makes very little sense! I'm definitely fine with just raising an issue that the `clip` dtype resolution needs more work and keeping things as you have them here.